views/components/form-fields/macros.njk

```diff
@@ -214,6 +217,122 @@
     </fieldset>
 {% endmacro %}
 
+{% macro budgetLineItemRow(field, counter, lineItem = false) %}
+    <tr>
+        <td>
+            <input type="text"
+                   autocomplete="off"
+                   name="{{ field.name }}[{{ counter }}][item]"
+                   placeholder="Add item name"
+                   {% if lineItem %}value="{{ lineItem.item }}"{% endif %}
+                   class="ff-text u-block-full" />
+        </td>
+        <td>
+            <div class="ff-field-prepend">£</div>
+            <input type="number"
+                   name="{{ field.name }}[{{ counter }}][cost]"
+                   min="1" max="{{ field.attributes.max }}"
+                   {% if lineItem %}value="{{ lineItem.cost }}"{% endif %}
+                   class="ff-text ff-text--currency" />
+        </td>
+        <td class="js-only">
+            <button class="btn btn--small btn--outline u-block-full">
+                {{ iconBin() }}
+                Delete row
+            </button>
+        </td>
+    </tr>
+{% endmacro %}
+
+{% macro inputBudget(field, fieldCopy) %}
+    <label class="ff-label" for="field-{{ field.name }}">
+        {{ fieldLabelText(field, fieldCopy) }}
+    </label>
+
+    {{ fieldHelpText(field, fieldCopy) }}
+
+    <div class="ff-budget"
```

Comment:
This should really be a vue component rather than inline, we can then use props to pass data around rather that data attrs. Same approach as past grants.

Reply:
I started it like that but I wanted to approach this like a hybrid component as much as possible (eg. like the free merch material form) and keep the code together rather than having two lots of similar/related markup in both a Vue file and in this macro. Past grants uses a bunch of JSON embedded in `<script>` tags to populate state – I think anytime we're doing this kind of PE work with Vue, there's going to be tradeoffs. In this instance I wanted to keep all the related code together where possible.

Reply:
Given the number of things this component has to cover it would be nice if we could get it into a self contained vue component as I think that gives us some strong benefits when it comes to testing as we can run much faster unit-level tests on the component using the `vue-test-utils` stuff in addition to some coverage from Cypress. As it stands we'd have to cover everything just using Cypress which would be much slower. I might be missing something where we could get the best of both worlds though, so open to exploring a few things.

Reply:
Just converted this over to a self-contained component, give it a look!